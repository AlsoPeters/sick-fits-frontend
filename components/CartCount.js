export default function CartCount({ count }) {
  return (
    <div className="p-2 mx-2 text-base font-bold leading-4 rounded-full tabular-nums h-min min-w-fit bg-tokyo-term-black">
      {count}
    </div>
  );
}
