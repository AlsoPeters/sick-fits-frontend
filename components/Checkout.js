import { loadStripe } from '@stripe/stripe-js';
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from '@stripe/react-stripe-js';
import { useState } from 'react';
import nProgress from 'nprogress';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/router';
import { useCart } from '../lib/cartState';
import { CURRENT_USER_QUERY } from './User';

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const { closeCart } = useCart();
  const [checkout, { error: graphQLError }] = useMutation(
    CREATE_ORDER_MUTATION,
    {
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );

  async function handleSubmit(e) {
    // 1. Stop the form from submitting and turn the loader on
    e.preventDefault();
    setLoading(true);
    // 2. Start the page transition
    nProgress.start();
    // 3. Create the payment method via stripe (Token comes back here if successful)
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });
    console.log(paymentMethod);
    // 4. Handle any errors from stripe
    if (error) {
      setError(error);
      nProgress.done();
      return; // stops the checkout from happening
    }
    // 5. Send the token from step 3 to our keystone server, via a custom mutation
    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });
    console.log(`Finished with the order.`);
    console.log(order);
    // 6. Change the page to view the order
    router.push({
      pathname: `/order/[id]`,
      query: { id: order.data.checkout.id },
    });
    // 7. Close the cart
    closeCart();
    // 8. Turn the loader off
    setLoading(false);
    nProgress.done();
  }

  return (
    <form
      onSubmit={handleSubmit}
      className="w-auto p-4 mx-2 my-1 border-2 border-tokyo-term-magenta"
    >
      {error && <p className="text-xl text-tokyo-term-red">{error.message}</p>}
      {graphQLError && (
        <p className="text-xl text-tokyo-term-red">{graphQLError.message}</p>
      )}

      <p>hello</p>
      <CardElement />
      <button className="w-full px-2 py-1 text-base font-bold border-2 bg-tokyo-night_BLK border-tokyo-term-magenta">
        Check Out Now
      </button>
    </form>
  );
}

function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}

export { Checkout };
