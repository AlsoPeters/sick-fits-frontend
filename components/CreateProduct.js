import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import useForm from '../lib/useForm';
import DisplayError from '../components/ErrorMessage';
import { ALL_PRODUCTS_QUERY } from './Products';
import Router from 'next/router';

const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # Which variables are getting passed in? And what types are they?
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      price
      description
      name
    }
  }
`;

export default function CreateProduct() {
  const { inputs, handleChange, clearForm, resetForm } = useForm({
    image: '',
    name: 'Nice shoes',
    price: 1234,
    description: 'These are the best shoes!',
  });
  const [createProduct, { loading, error, data }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      variables: inputs,
      refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
    }
  );
  console.log(createProduct);

  return (
    <div>
      <form
        onSubmit={async (e) => {
          e.preventDefault();
          // Submit the inputfields to the backend:
          const res = await createProduct();
          clearForm();
          // Go to that products page
          Router.push({
            pathname: `/product/${res.data.createProduct.id}`,
          });
        }}
      >
        <DisplayError error={error} />
        <fieldset disabled={loading}>
          <label htmlFor="image">
            Image
            <input
              required
              type="file"
              id="image"
              name="image"
              onChange={handleChange}
            />
          </label>

          <label htmlFor="name">
            Name
            <input
              type="text"
              id="name"
              name="name"
              placeholder="Name"
              value={inputs.name}
              onChange={handleChange}
            />
          </label>
          <label htmlFor="price">
            price
            <input
              type="number"
              id="name"
              name="price"
              placeholder="Price"
              value={inputs.price}
              onChange={handleChange}
            />
          </label>
          <label htmlFor="description">
            Description
            <textarea
              id="description"
              name="description"
              placeholder="Description"
              value={inputs.description}
              onChange={handleChange}
            />
          </label>

          <button
            // TODO: add this button style as a default
            className="px-2 py-1 m-2 border-2 rounded-md bg-tokyo-term-black border-tokyo-term-magenta"
            type="submit"
          >
            + Add Product
          </button>
        </fieldset>
      </form>
    </div>
  );
}
