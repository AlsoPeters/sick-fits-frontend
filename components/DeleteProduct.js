import { useMutation } from "@apollo/client";
import gql from "graphql-tag";

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

function update(cache, payload) {
  console.log(payload);
  console.log("Running the update function after delete");
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export default function DeleteProduct({ id, children }) {
  const [deleteProduct, { loading, error }] = useMutation(
    DELETE_PRODUCT_MUTATION,
    {
      variables: { id: id },
      update: update,
    }
  );
  return (
    <button
      disabled={loading}
      className="px-2 py-1 font-bold border-2 rounded-md bg-tokyo-term-black border-tokyo-term-magenta"
      type="button"
      onClick={() => {
        if (confirm("Are you sure you want to delete this item?")) {
          console.log("DELETING");
          deleteProduct().catch((err) => alert(err.message));
        }
      }}
    >
      {children}
    </button>
  );
}
