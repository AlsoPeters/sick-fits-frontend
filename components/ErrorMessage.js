import React from 'react';

import PropTypes from 'prop-types';

// const ErrorStyles = styled.div`
//   padding: 2rem;
//   background: white;
//   margin: 2rem 0;
//   border: 1px solid rgba(0, 0, 0, 0.05);
//   border-left: 5px solid red;
//   p {
//     margin: 0;
//     font-weight: 100;
//   }
//   strong {
//     margin-right: 1rem;
//   }
// `;

const DisplayError = ({ error }) => {
  if (!error || !error.message) return null;
  if (
    error.networkError &&
    error.networkError.result &&
    error.networkError.result.errors.length
  ) {
    return error.networkError.result.errors.map((error, i) => (
      <div
        className='p-4 m-4 border-2 rounded-md bg-tokyo-term-black text-tokyo-term-red border-tokyo-term-red'
        key={i}
      >
        <p className='font-bold' data-test='graphql-error'>
          <strong className='mr-2 text-2xl'>Shoot!</strong>
          {error.message.replace('GraphQL error: ', '')}
        </p>
      </div>
    ));
  }
  return (
    <div>
      <p data-test='graphql-error'>
        <strong className='m-4 rounded-md text-tokyo-term-red border-tokyo-term-red'>
          Shoot!
        </strong>
        {error.message.replace('GraphQL error: ', '')}
      </p>
    </div>
  );
};

DisplayError.defaultProps = {
  error: {},
};

DisplayError.propTypes = {
  error: PropTypes.object,
};

export default DisplayError;
