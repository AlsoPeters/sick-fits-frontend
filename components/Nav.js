import Link from 'next/link';
import { useCart } from '../lib/cartState';
import CartCount from './CartCount';
import SignOut from './SignOut';
import { useUser } from './User';

export default function Nav() {
  const user = useUser();
  const { openCart } = useCart();
  return (
    <nav className='flex mx-4 w-max text-3xl font-bold'>
      <Link href='/products'>
        <p className='py-4 px-6 mx-2 rounded-md border-2 border-tokyo-term-magenta hover:cursor-pointer'>
          Products
        </p>
      </Link>
      {user && (
        <>
          <Link href='/sell'>
            <p className='py-4 px-6 mx-2 rounded-md border-2 border-tokyo-term-magenta hover:cursor-pointer'>
              Sell
            </p>
          </Link>
          <Link href='/orders'>
            <p className='py-4 px-6 mx-2 rounded-md border-2 border-tokyo-term-magenta hover:cursor-pointer'>
              Orders
            </p>
          </Link>
          <Link href='/account'>
            <p className='py-4 px-6 mx-2 rounded-md border-2 border-tokyo-term-magenta hover:cursor-pointer'>
              Account
            </p>
          </Link>
          <SignOut />
          <div className='flex items-center'>
            <button
              className='flex py-4 px-6 mx-2 rounded-md border-2 border-tokyo-term-magenta hover:cursor-pointer'
              type='button'
              onClick={openCart}
            >
              My Cart
              <CartCount
                count={user.cart.reduce(
                  (tally, cartItem) =>
                    tally + (cartItem.product ? cartItem.quantity : 0),
                  0
                )}
              />
            </button>
          </div>
        </>
      )}
      {!user && (
        <>
          <Link href='/signin'>
            <p className='py-4 px-6 mx-2 rounded-md border-2 border-tokyo-term-magenta hover:cursor-pointer'>
              Sign In
            </p>
          </Link>
        </>
      )}
    </nav>
  );
}
