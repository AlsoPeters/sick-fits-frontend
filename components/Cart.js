import { useCart } from '../lib/cartState';
import { useUser } from './User';
import formatMoney from '../lib/formatMoney';
import calcTotalPrice from '../lib/calcTotalPrice';
import RemoveFromCart from './RemoveFromCart';
import Checkout from './Checkout';

function CartItem({ cartItem }) {
  const product = cartItem.product;
  if (!product) return null;
  return (
    <li className='grid grid-cols-2 px-2 py-1 my-4 border-b border-tokyo-term-magenta'>
      <img
        className='w-full col-span-1 mr-4'
        src={product.photo.image.publicUrlTransformed}
        alt={product.name}
      />
      <div className='mx-2'>
        <h3 className='text-2xl font-bold'>{product.name}</h3>
        <p className='col-start-2 font-bold'>
          {formatMoney(product.price * cartItem.quantity)} -
          <em>
            {cartItem.quantity} &times; {formatMoney(product.price)} each
          </em>
        </p>
      </div>
      <div className='col-start-3'>
        <RemoveFromCart id={cartItem.id} />
      </div>
    </li>
  );
}

export default function Cart() {
  const me = useUser();
  const { cartOpen, closeCart } = useCart();
  if (!me) return null;
  return (
    <div className='mx-4 bg-tokyo-storm_BLK'>
      <header className='px-2 py-1 m-2 text-2xl font-bold border-2 rounded-sm bg-tokyo-term-black border-tokyo-term-magenta w-max'>
        {me.name}'s Cart
      </header>
      <button
        className='px-2 py-1 text-xl border-2 rounded-md border-tokyo-term-magenta bg-tokyo-term-black'
        onClick={closeCart}
      >
        &times;
      </button>
      <ul>
        {me.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer className='flex flex-col w-full px-2 mx-2 text-3xl font-bold text-tokyo-term-green bg-tokyo-term-black'>
        {formatMoney(calcTotalPrice(me.cart))}
        <Checkout />
      </footer>
      {/* // TODO: Make slide in transition */}
    </div>
  );
}
