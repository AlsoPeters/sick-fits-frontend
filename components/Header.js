import Link from 'next/link';
import Cart from './Cart';
import Nav from './Nav';
import Search from './Search';

export default function Header() {
  return (
    <header className='bg-tokyo-night_BLK'>
      <div className='grid grid-cols-2 items-center'>
        <Link href='/'>
          <h1 className='m-4 logo-primary hover:cursor-pointer'>Sick Fits</h1>
        </Link>
        <div className='grid place-content-end'>
          <Nav />
        </div>
      </div>
      <div className='ml-4'>
        <Search />
      </div>
      <Cart />
    </header>
  );
}
