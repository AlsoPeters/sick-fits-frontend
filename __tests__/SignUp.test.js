import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import Signup, { SIGNUP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const me = fakeUser();
const password = 'kyle';

const mocks = [
  // Mutation Mock
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password: password,
      },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  // Current user mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: me } },
  },
];

describe('<SignUp/>', () => {
  it('Renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <Signup />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it('Calls the mutation properly', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <Signup />
      </MockedProvider>
    );
    // Type into the boxes
    await userEvent.type(screen.getByPlaceholderText('Dohn Joe'), me.name);
    await userEvent.type(
      screen.getByPlaceholderText('Your Email Adress'),
      me.email
    );
    await userEvent.type(screen.getByPlaceholderText('Password'), password);
    // Click the submit
    await userEvent.click(screen.getByText('Sign Up'));
    await screen.findByText(
      `Signed up with ${me.email} - Please go head and sign in.`
    );
    debug();
  });
});
