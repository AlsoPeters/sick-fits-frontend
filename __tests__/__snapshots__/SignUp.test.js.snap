// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> Renders and matches snapshot 1`] = `
<div>
  <form
    class="flex"
    method="POST"
  >
    <fieldset
      class="px-2 border-2 rounded-sm border-tokyo-term-magenta"
    >
      <h2
        class="text-xl font-bold text-tokyo-term-white "
      >
        Sign Up For an Account
      </h2>
      <label
        class="flex flex-col my-2 font-bold"
        for="email"
      >
        Your Name
        <input
          autocomplete="name"
          class="px-2 border rounded-sm text-tokyo-term-white bg-tokyo-comment_PURP border-tokyo-term-magenta"
          name="name"
          placeholder="Dohn Joe"
          type="text"
          value=""
        />
      </label>
      <label
        class="flex flex-col my-2 font-bold"
        for="email"
      >
        Email
        <input
          autocomplete="email"
          class="px-2 border rounded-sm text-tokyo-term-white bg-tokyo-comment_PURP border-tokyo-term-magenta"
          name="email"
          placeholder="Your Email Adress"
          type="email"
          value=""
        />
      </label>
      <label
        class="flex flex-col my-2 font-bold"
        for="password"
      >
        Password
        <input
          autocomplete="password"
          class="px-2 border rounded-sm text-tokyo-term-white bg-tokyo-comment_PURP border-tokyo-term-magenta"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        class="px-2 my-2 font-bold rounded-sm bg-tokyo-term-magenta text-tokyo-night_BLK"
        type="submit"
      >
        Sign Up
      </button>
    </fieldset>
  </form>
</div>
`;
